### PR TITLE
feat: Add onboarding and development setup for Max Battle

### LINKED ISSUE #2: Initial developer onboarding for Max Battle
# Developer Setup

:wave: Hi @maxematical. This first Issue is meant to help get your work computer into a state where it is ready for software development and have you document how you got there for your future self and others. This will also serve as a short exercise in making a PR and communication. Depending on your familiarity level with the tools covered this may take you a day or a week &mdash; both are fine, just please remember to track and record the time that you spend working.

There is no need to treat this like homework where you have to try to figure things out by yourself. You can ask for help from me and others. The goal is for you to learn how to setup an environment for development and debugging before we move onto other development tasks, so the steps are intentionally meant to be high level directives that you will need to figure out how to implement (they will vary drastically in time required).

## Initial setup

1. Install Git on your work machine.

2. Read [How to Write a Git Commit Message](https://cbea.ms/git-commit/).

3. Work through The Carpentries module [Version Control with Git](https://swcarpentry.github.io/git-novice/).

4. Read over GitHub's overview of [GitHub flow](https://docs.github.com/en/get-started/quickstart/github-flow).

5. **If your local machine's operating system is Windows** install Windows Subsystem for Linux (WSL) 2 on it.

6. Ensure that you have a working package manger installed.
   * If your local machine is running macOS I would recommend [Homebrew](https://brew.sh/).
   * If your local machine is running a Linux distribution there should be a built in package manager.
   * If your local machine is running Windows, use the `apt` package manager in the Ubuntu environment from WSL2.

7. (Optional) If you would like some additional practice with Git in a simulated file system, visit [Learn Git Branching](https://learngitbranching.js.org/) and work through the exercises there.

## Steps

Make sure that you document each of these steps.

1. Make a public fork of this repository to your personal GitHub account.

2. Generate SSH keys and add them to your GitHub account.

3. Using your SSH keys clone your fork to your machine and create a feature branch where you will do this work.

4. Create a directory in the top level of the repository that is your last name. Create a file inside of that named `README.md` and continue to document all of these steps in this Markdown file.

5. Create a draft pull request of these changes from your fork to the main project. As you proceed through the rest of this work continue to make regular commits and push your changes. Use a GitHub keyword in the body of the PR description to automatically close this Issue when the PR is merged.

6. Select and install an IDE to your local machine. If you do not know or care what IDE to use, select VS Code. You are not required to use this IDE for the rest of the exercise if you would prefer to only use a text editor, but use of an IDE is a good thing to know at some level.

7. Choose your favorite programming language (it can be any language) and write a small example program in it. It can be a "hello world" or an example of something that you like about the language you selected. Add the source files, required build system files, and instructions on how to build and run the program to your PR.

8. Install [pre-commit](https://pre-commit.com/) and add a pre-commit config file to your PR.

9. Install Docker on your work machine.

10. Work through [The Carpentries Incubator module on Docker](https://carpentries-incubator.github.io/docker-introduction/).

12. Write a Dockerfile that when built will build and run your earlier example program. Include this Dockerfile and build instructions or script in your PR.

13. Move your PR out of draft mode and tag Matthew to review the PR.

### PR DESCRIPTION
**Pull request by maxematical**
(Closes AFIDSI/student-developer-setup#2.)

This is my initial developer setup "project" for working at DSI -- mainly installing some of the tools that are used and getting some practice with using pull requests/Git/Docker.

This pull request contains a "hello world" program written in C that you can run using a Docker container. You can set up (build) a docker image that compiles and runs the program by going into the `battle` folder and running `docker build -t my_image_name .`. Then run the image via `docker container run --rm my_image_name`. You should see it print out `Hello world!` !

### CODE REVIEW

**Comment by maxematical** [comment]
@matthewfeickert Hey Matthew, I was able to finish all the instructions and set everything up. Let me know if you have any feedback, thanks!

File: .pre-commit-config.yaml
Comment (matthewfeickert) [inline .pre-commit-config.yaml:1]: This is fine, but can you move this under your `battle` directory? That was we don't get merge conflicts from everyone else adding their `pre-commit` configs in their PRs.

File: battle/README.md
Comment (matthewfeickert) [inline battle/README.md:5]: This is good and pretty close to what I was going for. The main reason for documenting things is to have you get in the practice of having to describe to others what you did and why when you're communicating about open source. Slightly more detail here (not much) would be good to include. For example, how did you install Git? Did you use a package manager? Did you manually install it? Did you build it from source? What distribution method did you get it from?

**Review by matthewfeickert — CHANGES_REQUESTED** [review]
@maxematical this is a good start. :+1: In addition to the review comments I've got a few pieces of general feedback.

In Issue #2 it mentioned

> Using your SSH keys clone your fork to your machine and create a feature branch where you will do this work.

This PR is coming from your `main` branch. In general, you don't want to develop directly on a project's default branch (for most that will be `main` but any branch can be set as the default branch). Instead, you want to always create a feature branch where you can do all of your work as your PR progresses. There are many reasons for this, but a few are:

* Developing on a default branch runs the risk of making it impossible to rebase the branch without accidentally closing the PR.
* It means that you aren't going to be able to merge in other changes that happen on the upstream project's default branch _and_ have the same history as theirs unless you do some fancy `git` surgery. This makes it harder and slower to contribute if the project has a lot of "velocity" behind it and is accepting several PRs a day.
* Commits to the default branch that are pushed to a Git hosting service (i.e. GitHub) usually trigger additional continuous integration (CI) automated workflows while committing to an arbitrarily named feature branch will not. Running these extra workflows might have unintended side effects and probably just wastes CPU cycles.
* Different projects have different ways of accepting contributions. Some of them will directly merge your PR, but others will do a "squash and merge" workflow that "squash" all your commits into one single one and then just merge that &mdash; I use this for most of my projects as I find that it is a good way to incentivize "atomic" commits.
* There's more but I've already gone on too much.

> Write a Dockerfile that when built will build and run your earlier example program. Include this Dockerfile and build instructions or script in your PR.

You added your Dockerfile :+1: but you need to add build instructions / script on how to build the Dockerfile into an image and then how to run the image.

You did a pretty good job with the PR title as well &mdash; a short description which gives an overview, though I find it useful to structure it in a similar way to how you would a [good commit message](https://cbea.ms/git-commit/) &mdash; something short that tells you what it will do if you apply it / accept it. Also in general you should add some short text to the PR body (doesn't have to be long) that explains to the maintainer of the project what this PR does, why you're doing it, what Issues it might address (you did that part here, which is good). So you should add just a few sentences here as well. :+1: 

Not a requested change, but as you're using VS Code have you found some of their plugins that are useful? If not, I can suggest some that you might want to start using.

Overall a good PR draft so far though!

File: .pre-commit-config.yaml
Comment (maxematical) [reply .pre-commit-config.yaml:1]: I tried that but if `.pre-commit-config.yaml` isn't in the "root" directory, it seems like `pre-commit` can't find the configuration file? When I have my config under the battle folder, `git commit` gives me this message:

```
No .pre-commit-config.yaml file was found
- To temporarily silence this, run `PRE_COMMIT_ALLOW_NO_CONFIG=1 git ...`
- To permanently silence this, install pre-commit with the --allow-missing-config option
- To uninstall pre-commit run `pre-commit uninstall`
```

File: battle/README.md
Comment (maxematical) [reply battle/README.md:5]: Ah that makes sense, thanks. I updated the checklist with more detailed notes for some of the steps now.

**Comment by maxematical** [comment]
@matthewfeickert Thanks for the feedback. I went in and added some more detail in the readme based on your comments and expanded the description of this pull request with more context.

For the instructions on how to use the Dockerfile, is it fine if I just put it in the PR description or do you want me to add it somewhere in the repository as well?

Also, I must have missed the step where it said to create a new branch and do the development there. Should I try to move the commits of this PR onto a new branch or is there not really a way to do that once the PR is created?

For VS Code plugins right now, I'm using the WSL plugin (so I can edit files in directly in the linux filesystem), the python plugin, and a markdown preview plugin. But any other ideas would be nice too!

File: .pre-commit-config.yaml
Comment (matthewfeickert) [reply .pre-commit-config.yaml:1]: Ah right. My bad, I had forgotten that this is by design &mdash; when I was creating these Issues I was thinking of each of them being a separate project and forgetting that they would all need to have to contend with merge conflicts.

Uh, since I didn't about this well enough in advance just leave it how it is and then I'll give everyone else advice about how to just merge in the results of whatever PR gets merged first. So no need to do anything here. :+1:

File: battle/README.md
Comment (matthewfeickert) [inline battle/README.md:74]: Responding inline here:

> For the instructions on how to use the Dockerfile, is it fine if I just put it in the PR description or do you want me to add it somewhere in the repository as well?

I see that you've added info in the PR body (which is great!) but it would be a good idea to also have the build and run commands for the `Dockerfile` here in the README as well. :+1: 

Other than that LGTM.

**Review by matthewfeickert — CHANGES_REQUESTED** [review]
Thanks for these revisions, @maxematical &mdash; things are looking very good and I just have one very small revision request which for you will just be a copy-paste. Other than that everything LGTM! :+1: 

Responding to some of your other questions:

> Also, I must have missed the step where it said to create a new branch and do the development there. Should I try to move the commits of this PR onto a new branch or is there not really a way to do that once the PR is created?

No worries, and no, please don't worry about trying to change the PR branch as that's more trouble than it is worth &mdash; just something to think about in the future.

Once this PR gets merged then I can give you some instructions on how to resync your fork with the "upstream" main project.

> For VS Code plugins right now, I'm using the WSL plugin (so I can edit files in directly in the linux filesystem), the python plugin, and a markdown preview plugin. But any other ideas would be nice too!

Sounds good. The good news/bad news about VS Code is there are a plethora of plugins and so you can become swamped with them quickly. I'll hold off on suggesting any others at the moment but once we get you working on a dedicated project I can make suggestions depending on your workflow.

File: .pre-commit-config.yaml
Comment (maxematical) [reply .pre-commit-config.yaml:1]: 👍

**Comment by maxematical** [comment]
@matthewfeickert thanks for the feedback again, I updated the README with the changes so let me know when you have a chance to look at it again!

File: battle/README.md
Comment (matthewfeickert) [inline battle/README.md:82]: This is fine, but the `container` bit isn't actually needed. You can just do

```
docker run --rm my_image_name
```

and (I think you know this but for completeness of my example) if you wanted an interactive session you can do

```console
$ docker run --rm -ti my_image_name /bin/bash
root@f3b44bcfca88:/home# ls
build.sh  hello  hello.c
root@f3b44bcfca88:/home# ./hello Moby 
Hello, Moby
root@f3b44bcfca88:/home#
```

**Review by matthewfeickert — APPROVED** [review]
Thanks for this PR @maxematical. LGTM! :shipit:

I'm going to merge this in and post instructions to everyone else on how to deal with the existence of the `.pre-commit-config.yaml` and then I'll follow up with you.

**Comment by matthewfeickert** [comment]
@maxematical Okay what I did when I merged this PR in is called a "squash and merge". On `main` all of your work on this branch is now "squashed" down into a single commit. As you worked on your `main` branch you now have a conflict between your local `main` and the "upstream" `main` of https://github.com/AFIDSI/student-developer-setup/tree/main. To get around this we're to use a power of Git &mdash; history rewriting (which some people call the chainsaw feature of Git, so we'll be careful).

The first thing you're going to do is update your local fork's information on all other versions of the Git repository

```
git checkout main
git fetch --all --prune
```

When you do this you'll see that your local `main` now has commits both ahead and behind of the upstream's `main`. You can visualize this easier at the command line if you first create a [Git alias](https://github.com/matthewfeickert/git-basics#helpful-aliases) by running

```
git config --global alias.glog 'log --graph --oneline --decorate --all'
```

and then run

```
git glog
```

which should show you the following in your terminal

```
* 7fadc2b (upstream/main) feat: Add onboarding and development setup for Max Battle (#11)
| * 8b043aa (HEAD -> main, origin/main, origin/HEAD) Add Dockerfile instructions to README, make formatting consistent
| * 01c5c7f Update README with more detailed notes and ignore binaries
| * 64e86a5 Update README with completed steps and information about hello-world program
| * ca0ed1a Create a script to build the program and only copy needed files to docker image
| * efbca1e Convert hello-world program to C, create dockerfile to compile and run the C program
| * 49d10c4 Move pre-commit configuration file to repository root
| * 7442ce3 Add hello world script and update battle/README.md
| * 08ce731 Create battle/README.md with information about currently completed steps
|/  
* 9a387d8 chore: Remove line breaks from onboarding template (#6)
* 38a08d5 feat: Add developer onboarding Issue template (#1)
* bd18d18 Initial commit
```

What you'll do now is tell Git that you forcibly want to overwrite your `main` branches history with the history of the upstream's `main`. You basically want to avoid doing this in general, but we're going to do it here. You can do this with

```
git reset --hard upstream/main
```

which will now reset your _local_ history to be the same as `upstream/main`. We still need to tell your fork about this new information, so we're going to again do a dangerous thing and force push to your remote (again overwriting history!).

```
git push --force-with-lease origin HEAD
```

When you do this you should then see that your local repo, your fork on GitHub, and the upstream on GitHub are now all in the same state and you could start development again. :+1: 

<details>
<summary>Aside on what happens when things go wrong:</summary>

You can imagine that sometimes this stuff goes wrong. The good news is that even if everything looks bleak Git gives you a way to undo anything bad. Git keeps track of _every_ command that you run and gives you a look up of it with [`git reflog`](https://git-scm.com/docs/git-reflog). If you run it at the command line

```
git reflog
```

it shows a complete history of Git commands for the repository. If things have gone super wrong just run it and fine the SHA (the 7 digit number on the right of each line) of the state of the repo that things were in before things went wrong. Then simply run

```
git reset --hard <sha>
```

and you'll be right back to where you were before things got messy. :+1: 

</details>

Now that you've completed your initial onboarding please email me or send me a Slack message so we can setup a meeting for next week to get you going on a project.

**Comment by matthewfeickert** [comment]
Mentioning @maxematical's contribution history Issue https://github.com/AFIDSI/open-source-contributions/issues/2 to kick things off.:+1:

**Pull request merged**
